File: lib/DevOther/src/main/java/dev/other/MMKVUtils.java
package dev.other;

import android.content.Context;
import android.os.Parcelable;

import com.tencent.mmkv.MMKV;
import com.tencent.mmkv.MMKVLogLevel;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import dev.utils.LogPrintUtils;
import dev.utils.common.StringUtils;

/**
 * detail: MMKV 工具类
 * @author Ttt
 * <pre>
 *     支持组件化 module 存储、以及默认通用 mmkv 对象
 *     <p></p>
 *     基于 mmap 的高性能通用 key-value 组件
 *     @see <a href="https://github.com/Tencent/MMKV/blob/master/readme_cn.md"/>
 *     [Google] 再见 SharedPreferences 拥抱 Jetpack DataStore
 *     @see <a href="https://juejin.im/post/6881442312560803853"/>
 *     [Google] 再见 SharedPreferences 拥抱 Jetpack DataStore ( 二 )
 *     @see <a href="https://juejin.im/post/6888847647802097672"/>
 *     <p></p>
 *     {@link MMKVUtils#defaultHolder()}.encode/decodeXxx
 *     {@link MMKVUtils#get(String)}.encode/decodeXxx
 * </pre>
 */
public final class MMKVUtils {

    private MMKVUtils() {
    }

    // 日志 TAG
    private static final String TAG = MMKVUtils.class.getSimpleName();

    // 持有类存储 Key-Holder
    private static final Map<String, Holder> HOLDER_MAP     = new HashMap<>();
    // Default MMKV Holder
    private static       Holder              DEFAULT_HOLDER = null;

    /**
     * 初始化方法 ( 必须调用 )
     * @param context {@link Context}
     */
    public static void init(final Context context) {
        // 初始化 MMKV 并设置日志级别
        String rootDir = MMKV.initialize(context, MMKVLogLevel.LevelNone);
        LogPrintUtils.dTag(TAG, "MMKV rootDir: %s", rootDir);

//        // 设置打印日志级别
//        MMKV.setLogLevel(MMKVLogLevel.LevelNone);

//        https://github.com/Tencent/MMKV/wiki/android_advance_cn
//        // 视项目需求进行注册监听、数据恢复等
//        MMKV.registerHandler();
    }

    // ===============
    // = 对外公开方法 =
    // ===============

    /**
     * 是否存在指定 Key 的 MMKV Holder
     * @param key Key
     * @return {@code true} yes, {@code false} no
     */
    public static boolean containsMMKV(final String key) {
        return HOLDER_MAP.containsKey(key);
    }

    /**
     * 通过 Key 获取 MMKV Holder
     * @param key Key
     * @return {@link Holder}
     */
    public static Holder get(final String key) {
        if (containsMMKV(key)) return HOLDER_MAP.get(key);
        return putHolder(key);
    }

    /**
     * 保存自定义 MMKV Holder
     * @param key Key
     * @return {@link Holder}
     */
    public static Holder putHolder(final String key) {
        return putHolder(key, MMKV.mmkvWithID(key, MMKV.MULTI_PROCESS_MODE));
    }

    /**
     * 保存自定义 MMKV Holder
     * @param key  Key
     * @param mmkv {@link MMKV}
     * @return {@link Holder}
     */
    public static Holder putHolder(
            final String key,
            final MMKV mmkv
    ) {
        Holder holder = new Holder(mmkv);
        HOLDER_MAP.put(key, holder);
        return holder;
    }

    // =

    /**
     * 获取 Default MMKV Holder
     * @return {@link Holder}
     */
    public static Holder defaultHolder() {
        if (DEFAULT_HOLDER == null) {
            MMKV mmkv = MMKV.mmkvWithID(TAG, MMKV.MULTI_PROCESS_MODE);
            DEFAULT_HOLDER = new Holder(mmkv);
        }
        return DEFAULT_HOLDER;
    }

    // =============
    // = 内部封装类 =
    // =============

    /**
     * detail: MMKV 持有类
     * @author Ttt
     * <pre>
     *     提供常用方法, 可根据需求自行添加修改或通过 {@link #getMMKV()} 进行操作
     * </pre>
     */
    public static final class Holder {

        // MMKV
        private final MMKV mmkv;

        public Holder(MMKV mmkv) {
            this.mmkv = mmkv;
        }

        /**
         * 获取 MMKV
         * @return {@link MMKV}
         */
        public MMKV getMMKV() {
            return mmkv;
        }

        /**
         * 获取 MMKV mmap id
         * @return mmap id
         */
        public String mmapID() {
            if (isMMKVEmpty()) return null;
            return mmkv.mmapID();
        }

        // ===========
        // = 其他操作 =
        // ===========

        /**
         * 判断 MMKV 是否为 null
         * @return {@code true} yes, {@code false} no
         */
        public boolean isMMKVEmpty() {
            return mmkv == null;
        }

        /**
         * 判断 MMKV 是否不为 null
         * @return {@code true} yes, {@code false} no
         */
        public boolean isMMKVNotEmpty() {
            return mmkv != null;
        }

        /**
         * 是否存在指定 Key value
         * @param key Key
         * @return {@code true} yes, {@code false} no
         */
        public boolean containsKey(String key) {
            if (isMMKVEmpty()) return false;
            if (StringUtils.isEmpty(key)) return false;
            return mmkv.containsKey(key);
        }

        /**
         * 通过 key 移除 value
         * @param key Key
         * @return {@code true} success, {@code false} fail
         */
        public boolean removeValueForKey(String key) {
            if (isMMKVEmpty()) return false;
            if (StringUtils.isEmpty(key)) return false;
            mmkv.removeValueForKey(key);
            return true;
        }

        /**
         * 通过 key 数组移除 value
         * @param keys key 数组
         * @return {@code true} success, {@code false} fail
         */
        public boolean removeValuesForKeys(String[] keys) {
            if (isMMKVEmpty()) return false;
            if (keys == null) return false;
            mmkv.removeValuesForKeys(keys);
            return true;
        }

        /**
         * 同步操作
         * @return {@code true} success, {@code false} fail
         */
        public boolean sync() {
            if (isMMKVEmpty()) return false;
            mmkv.sync();
            return true;
        }

        /**
         * 异步操作
         * @return {@code true} success, {@code false} fail
         */
        public boolean async() {
            if (isMMKVEmpty()) return false;
            mmkv.async();
            return true;
        }

        // =======
        // = 存储 =
        // =======

        public boolean encode(
                String key,
                boolean value
        ) {
            if (isMMKVEmpty()) return false;
            if (StringUtils.isEmpty(key)) return false;
            return mmkv.encode(key, value);
        }

        public boolean encode(
                String key,
                int value
        ) {
            if (isMMKVEmpty()) return false;
            if (StringUtils.isEmpty(key)) return false;
            return mmkv.encode(key, value);
        }

        public boolean encode(
                String key,
                long value
        ) {
            if (isMMKVEmpty()) return false;
            if (StringUtils.isEmpty(key)) return false;
            return mmkv.encode(key, value);
        }

        public boolean encode(
                String key,
                float value
        ) {
            if (isMMKVEmpty()) return false;
            if (StringUtils.isEmpty(key)) return false;
            return mmkv.encode(key, value);
        }

        public boolean encode(
                String key,
                double value
        ) {
            if (isMMKVEmpty()) return false;
            if (StringUtils.isEmpty(key)) return false;
            return mmkv.encode(key, value);
        }

        public boolean encode(
                String key,
                String value
        ) {
            if (isMMKVEmpty()) return false;
            if (StringUtils.isEmpty(key)) return false;
            return mmkv.encode(key, value);
        }

        public boolean encode(
                String key,
                Set<String> value
        ) {
            if (isMMKVEmpty()) return false;
            if (StringUtils.isEmpty(key)) return false;
            if (value == null) return false;
            return mmkv.encode(key, value);
        }

        public boolean encode(
                String key,
                byte[] value
        ) {
            if (isMMKVEmpty()) return false;
            if (StringUtils.isEmpty(key)) return false;
            if (value == null) return false;
            return mmkv.encode(key, value);
        }

        public boolean encode(
                String key,
                Parcelable value
        ) {
            if (isMMKVEmpty()) return false;
            if (StringUtils.isEmpty(key)) return false;
            if (value == null) return false;
            return mmkv.encode(key, value);
        }

        // =======
        // = 读取 =
        // =======

        public boolean decodeBool(String key) {
            return decodeBool(key, false);
        }

        public boolean decodeBool(
                String key,
                boolean defaultValue
        ) {
            if (isMMKVEmpty()) return defaultValue;
            if (StringUtils.isEmpty(key)) return defaultValue;
            return mmkv.decodeBool(key, defaultValue);
        }

        public int decodeInt(String key) {
            return decodeInt(key, 0);
        }

        public int decodeInt(
                String key,
                int defaultValue
        ) {
            if (isMMKVEmpty()) return defaultValue;
            if (StringUtils.isEmpty(key)) return defaultValue;
            return mmkv.decodeInt(key, defaultValue);
        }

        public long decodeLong(String key) {
            return decodeLong(key, 0L);
        }

        public long decodeLong(
                String key,
                long defaultValue
        ) {
            if (isMMKVEmpty()) return defaultValue;
            if (StringUtils.isEmpty(key)) return defaultValue;
            return mmkv.decodeLong(key, defaultValue);
        }

        public float decodeFloat(String key) {
            return decodeFloat(key, 0.0F);
        }

        public float decodeFloat(
                String key,
                float defaultValue
        ) {
            if (isMMKVEmpty()) return defaultValue;
            if (StringUtils.isEmpty(key)) return defaultValue;
            return mmkv.decodeFloat(key, defaultValue);
        }

        public double decodeDouble(String key) {
            return decodeDouble(key, 0.0D);
        }

        public double decodeDouble(
                String key,
                double defaultValue
        ) {
            if (isMMKVEmpty()) return defaultValue;
            if (StringUtils.isEmpty(key)) return defaultValue;
            return mmkv.decodeDouble(key, defaultValue);
        }

        public String decodeString(String key) {
            return decodeString(key, null);
        }

        public String decodeString(
                String key,
                String defaultValue
        ) {
            if (isMMKVEmpty()) return defaultValue;
            if (StringUtils.isEmpty(key)) return defaultValue;
            return mmkv.decodeString(key, defaultValue);
        }

        public Set<String> decodeStringSet(String key) {
            return decodeStringSet(key, null);
        }

        public Set<String> decodeStringSet(
                String key,
                Set<String> defaultValue
        ) {
            return decodeStringSet(key, defaultValue, HashSet.class);
        }

        public Set<String> decodeStringSet(
                String key,
                Set<String> defaultValue,
                Class<? extends Set> cls
        ) {
            if (isMMKVEmpty()) return defaultValue;
            if (StringUtils.isEmpty(key)) return defaultValue;
            return mmkv.decodeStringSet(key, defaultValue, cls);
        }

        public byte[] decodeBytes(String key) {
            return decodeBytes(key, null);
        }

        public byte[] decodeBytes(
                String key,
                byte[] defaultValue
        ) {
            if (isMMKVEmpty()) return defaultValue;
            if (StringUtils.isEmpty(key)) return defaultValue;
            return mmkv.decodeBytes(key, defaultValue);
        }

        public <T extends Parcelable> T decodeParcelable(
                String key,
                Class<T> tClass
        ) {
            return decodeParcelable(key, tClass, null);
        }

        public <T extends Parcelable> T decodeParcelable(
                String key,
                Class<T> tClass,
                T defaultValue
        ) {
            if (isMMKVEmpty()) return defaultValue;
            if (StringUtils.isEmpty(key)) return defaultValue;
            return mmkv.decodeParcelable(key, tClass, defaultValue);
        }
    }
}

File: lib/DevEngine/src/main/java/dev/engine/keyvalue/MMKVUtils.kt
package dev.engine.keyvalue

import android.content.Context
import android.os.Parcelable
import com.tencent.mmkv.MMKV
import com.tencent.mmkv.MMKVLogLevel
import dev.engine.keyvalue.MMKVUtils.defaultHolder
import dev.engine.keyvalue.MMKVUtils.get
import dev.utils.LogPrintUtils
import dev.utils.common.StringUtils
import java.util.*
import kotlin.collections.set

/**
 * 初始化方法 ( 必须调用 )
 * @param context [Context]
 */
fun mmkvInitialize(context: Context?) {
    // 初始化 MMKV 并设置日志级别
    MMKV.initialize(context, MMKVLogLevel.LevelNone)
}

/**
 * detail: MMKV 工具类
 * @author Ttt
 * 支持组件化 module 存储、以及默认通用 mmkv 对象
 * 基于 mmap 的高性能通用 key-value 组件
 * @see https://github.com/Tencent/MMKV/blob/master/readme_cn.md
 * Google 再见 SharedPreferences 拥抱 Jetpack DataStore
 * @see https://juejin.im/post/6881442312560803853
 * Google 再见 SharedPreferences 拥抱 Jetpack DataStore
 * @see https://juejin.im/post/6888847647802097672
 * 使用:
 * [defaultHolder].encode/decodeXxx
 * [get].encode/decodeXxx
 */
internal object MMKVUtils {

    // 日志 TAG
    private val TAG = MMKVUtils::class.java.simpleName

    // 持有类存储 Key-Holder
    private val HOLDER_MAP: MutableMap<String, Holder> = HashMap()

    // Default MMKV Holder
    private val DEFAULT_HOLDER: Holder by lazy {
        val mmkv = MMKV.mmkvWithID(TAG, MMKV.MULTI_PROCESS_MODE)
        Holder(mmkv)
    }

    /**
     * 初始化方法 ( 必须调用 )
     * @param context [Context]
     */
    fun initialize(context: Context?) {
        // 初始化 MMKV 并设置日志级别
        val rootDir = MMKV.initialize(context, MMKVLogLevel.LevelNone)
        LogPrintUtils.dTag(TAG, "MMKV rootDir: %s", rootDir)

//        // 设置打印日志级别
//        MMKV.setLogLevel(MMKVLogLevel.LevelNone)

//        https://github.com/Tencent/MMKV/wiki/android_advance_cn
//        // 视项目需求进行注册监听、数据恢复等
//        MMKV.registerHandler()
    }

    // =============
    // = 对外公开方法 =
    // =============

    /**
     * 是否存在指定 Key 的 MMKV Holder
     * @param key Key
     * @return `true` yes, `false` no
     */
    fun containsMMKV(key: String): Boolean {
        return HOLDER_MAP.containsKey(key)
    }

    /**
     * 通过 Key 获取 MMKV Holder
     * @param key Key
     * @return [Holder]
     */
    operator fun get(key: String): Holder {
        return if (containsMMKV(key)) HOLDER_MAP[key]!! else putHolder(key)
    }

    /**
     * 保存自定义 MMKV Holder
     * @param key Key
     * @return [Holder]
     */
    fun putHolder(key: String): Holder {
        return putHolder(key, MMKV.mmkvWithID(key, MMKV.MULTI_PROCESS_MODE))
    }

    /**
     * 保存自定义 MMKV Holder
     * @param key  Key
     * @param mmkv [MMKV]
     * @return [Holder]
     */
    fun putHolder(
        key: String,
        mmkv: MMKV?
    ): Holder {
        val holder = Holder(mmkv)
        HOLDER_MAP[key] = holder
        return holder
    }

    // =

    /**
     * 获取 Default MMKV Holder
     * @return [Holder]
     */
    fun defaultHolder(): Holder {
        return DEFAULT_HOLDER
    }

    // ============
    // = 内部封装类 =
    // ============

    /**
     * detail: MMKV 持有类
     * @author Ttt
     * 提供常用方法, 可根据需求自行添加修改或通过 [Holder].mmkv 进行操作
     */
    class Holder(
        // MMKV
        val mmkv: MMKV?
    ) {

        /**
         * 获取 MMKV mmap id
         * @return mmap id
         */
        fun mmapID(): String? {
            return mmkv?.mmapID()
        }

        // ==========
        // = 其他操作 =
        // ==========

        /**
         * 判断 MMKV 是否为 null
         * @return `true` yes, `false` no
         */
        fun isMMKVEmpty(): Boolean {
            return mmkv == null
        }

        /**
         * 判断 MMKV 是否不为 null
         * @return `true` yes, `false` no
         */
        fun isMMKVNotEmpty(): Boolean {
            return mmkv != null
        }

        /**
         * 是否存在指定 Key value
         * @param key Key
         * @return `true` yes, `false` no
         */
        fun containsKey(key: String?): Boolean {
            if (StringUtils.isEmpty(key)) return false
            return mmkv?.containsKey(key) ?: false
        }

        /**
         * 通过 key 移除 value
         * @param key Key
         * @return `true` success, `false` fail
         */
        fun removeValueForKey(key: String?): Boolean {
            if (isMMKVEmpty()) return false
            if (StringUtils.isEmpty(key)) return false
            mmkv?.removeValueForKey(key)
            return true
        }

        /**
         * 通过 key 数组移除 value
         * @param keys key 数组
         * @return `true` success, `false` fail
         */
        fun removeValuesForKeys(keys: Array<String?>?): Boolean {
            if (isMMKVEmpty()) return false
            if (keys == null) return false
            mmkv?.removeValuesForKeys(keys)
            return true
        }

        /**
         * 同步操作
         * @return `true` success, `false` fail
         */
        fun sync(): Boolean {
            mmkv?.let {
                it.sync()
                return true
            }
            return false
        }

        /**
         * 异步操作
         * @return `true` success, `false` fail
         */
        fun async(): Boolean {
            mmkv?.let {
                it.async()
                return true
            }
            return false
        }

        /**
         * 清除全部数据
         * @return `true` success, `false` fail
         */
        fun clear(): Boolean {
            mmkv?.let {
                it.clear()
                return true
            }
            return false
        }

        // =======
        // = 存储 =
        // =======

        fun encode(
            key: String?,
            value: Boolean
        ): Boolean {
            if (StringUtils.isEmpty(key)) return false
            return mmkv?.encode(key, value) ?: false
        }

        fun encode(
            key: String?,
            value: Int
        ): Boolean {
            if (StringUtils.isEmpty(key)) return false
            return mmkv?.encode(key, value) ?: false
        }

        fun encode(
            key: String?,
            value: Long
        ): Boolean {
            if (StringUtils.isEmpty(key)) return false
            return mmkv?.encode(key, value) ?: false
        }

        fun encode(
            key: String?,
            value: Float
        ): Boolean {
            if (StringUtils.isEmpty(key)) return false
            return mmkv?.encode(key, value) ?: false
        }

        fun encode(
            key: String?,
            value: Double
        ): Boolean {
            if (StringUtils.isEmpty(key)) return false
            return mmkv?.encode(key, value) ?: false
        }

        fun encode(
            key: String?,
            value: String?
        ): Boolean {
            if (StringUtils.isEmpty(key)) return false
            return mmkv?.encode(key, value) ?: false
        }

        fun encode(
            key: String?,
            value: Set<String?>?
        ): Boolean {
            if (value == null) return false
            if (StringUtils.isEmpty(key)) return false
            return mmkv?.encode(key, value) ?: false
        }

        fun encode(
            key: String?,
            value: ByteArray?
        ): Boolean {
            if (value == null) return false
            if (StringUtils.isEmpty(key)) return false
            return mmkv?.encode(key, value) ?: false
        }

        fun encode(
            key: String?,
            value: Parcelable?
        ): Boolean {
            if (value == null) return false
            if (StringUtils.isEmpty(key)) return false
            return mmkv?.encode(key, value) ?: false
        }

        // =======
        // = 读取 =
        // =======

        fun decodeBool(key: String?): Boolean {
            return decodeBool(key, false)
        }

        fun decodeBool(
            key: String?,
            defaultValue: Boolean
        ): Boolean {
            if (StringUtils.isEmpty(key)) return defaultValue
            return mmkv?.decodeBool(key, defaultValue) ?: defaultValue
        }

        fun decodeInt(key: String?): Int {
            return decodeInt(key, 0)
        }

        fun decodeInt(
            key: String?,
            defaultValue: Int
        ): Int {
            if (StringUtils.isEmpty(key)) return defaultValue
            return mmkv?.decodeInt(key, defaultValue) ?: defaultValue
        }

        fun decodeLong(key: String?): Long {
            return decodeLong(key, 0L)
        }

        fun decodeLong(
            key: String?,
            defaultValue: Long
        ): Long {
            if (StringUtils.isEmpty(key)) return defaultValue
            return mmkv?.decodeLong(key, defaultValue) ?: defaultValue
        }

        fun decodeFloat(key: String?): Float {
            return decodeFloat(key, 0.0f)
        }

        fun decodeFloat(
            key: String?,
            defaultValue: Float
        ): Float {
            if (StringUtils.isEmpty(key)) return defaultValue
            return mmkv?.decodeFloat(key, defaultValue) ?: defaultValue
        }

        fun decodeDouble(key: String?): Double {
            return decodeDouble(key, 0.0)
        }

        fun decodeDouble(
            key: String?,
            defaultValue: Double
        ): Double {
            if (StringUtils.isEmpty(key)) return defaultValue
            return mmkv?.decodeDouble(key, defaultValue) ?: defaultValue
        }

        fun decodeString(key: String?): String? {
            return decodeString(key, null)
        }

        fun decodeString(
            key: String?,
            defaultValue: String?
        ): String? {
            if (StringUtils.isEmpty(key)) return defaultValue
            return mmkv?.decodeString(key, defaultValue) ?: defaultValue
        }

        fun decodeStringSet(key: String?): Set<String?>? {
            return decodeStringSet(key, null)
        }

        fun decodeStringSet(
            key: String?,
            defaultValue: Set<String?>?
        ): Set<String?>? {
            return decodeStringSet(key, defaultValue, HashSet::class.java)
        }

        fun decodeStringSet(
            key: String?,
            defaultValue: Set<String?>?,
            cls: Class<out Set<*>?>?
        ): Set<String?>? {
            if (StringUtils.isEmpty(key)) return defaultValue
            return mmkv?.decodeStringSet(key, defaultValue, cls) ?: defaultValue
        }

        fun decodeBytes(key: String?): ByteArray? {
            return decodeBytes(key, null)
        }

        fun decodeBytes(
            key: String?,
            defaultValue: ByteArray?
        ): ByteArray? {
            if (StringUtils.isEmpty(key)) return defaultValue
            return mmkv?.decodeBytes(key, defaultValue) ?: defaultValue
        }

        fun <T : Parcelable?> decodeParcelable(
            key: String?,
            tClass: Class<T>?
        ): T? {
            return decodeParcelable(key, tClass, null)
        }

        fun <T : Parcelable?> decodeParcelable(
            key: String?,
            tClass: Class<T>?,
            defaultValue: T?
        ): T? {
            if (StringUtils.isEmpty(key)) return defaultValue
            return mmkv?.decodeParcelable(key, tClass, defaultValue) ?: defaultValue
        }
    }
}